### ISSUE TITLE
Would you mind releasing the test code?

### ISSUE THREAD

**Issue by x-dragonet** [BradyFU/DVG#5]
Thank you for your hard work.  I can not reproduce the performance on the 1-fold of CASIA NIR-VIS 2.0 dataset after running your code, so I want to know your test code. This is what I proposed and I do look forward to receiving your reply at your convenience

**Comment by BradyFU** [comment]
Thank you for your attention to our work. However, we regret that it is inconvenent to release the testing code. For the released training code, we have run it over three times to make sure the accuracy. Thank you again for your understanding.

**Comment by AlfredXiangWu** [comment]
@x-dragonet 

It is easy to implement the testing for CASIA NIR-VIS 2.0. Following the protocols,
1. Get probe and gallery list.
2. Evaluate Rank-1 and ROC. 
  2.1 For rank-1 accuracy, you can select each probe image and compute the cosine similarity between it and all the gallery images. 
  2.2 For ROC, you can compute all the cosine similarity between probe and gallery (total about 200w+ pairs), and then feed them into a roc curve api such as sklearn.metrics.roc_curve. 

My testing code contains lots of other evaluation implementation, so it is not clear for CASIA NIR-VIS 2.0. When I am free, I will rewrite and release the testing code for CASIA NIR-VIS 2.0.

By the way, I think it is not difficult for you to implement a testing code and there are lots of reference abou ROC and rank-1 implementation on github. 

If any questions, feel free to concat with us.

**Comment by x-dragonet** [comment]
“nir_train_1.txt” and “vis_train_1.txt” only have 357 person. Why you set num_classes equal 725 in “run_train_lightcnn.sh” ?

**Comment by AlfredXiangWu** [comment]
CASIA NIR-VIS 2.0 contains totally about 725 identities and the testing protocol is divided into 10-fold. For convenience, we don't relabel the training list for each fold, which leads to discontinuous labels. 

From the perspective of softmax loss in deep embedding learning, each weight can be treated as the center of its corresponding identity. The weights of the 'empty' labels are initialized by Xavier and  cannot be updated in the training process. So they almost have no great influence for the embedding learning. 

In fact, we don't do any deep analysis for it. Maybe, the 'empty' label can be treated as a regularizer at the begining of the training process.

**Comment by x-dragonet** [comment]
Thank you, I also have some questions.
1.  your github provides the performance on the 1-fold of CASIA NIR-VIS 2.0 dataset.  The "train_list" on "run_train_generator.sh" is organized by “nir_train_1.txt” and “vis_train_1.txt” , but not use “nir_train_[2-10].txt” , “vis_train_[2-10].txt”, “nir_train_dev.txt”, "vis_train_dev.txt" . Is that right? 

2. In addition,  the performance on the 1-fold of CASIA NIR-VIS 2.0 dataset  is tested according to  
"nir_probe_1.txt" and "vis_gallery_1.txt",  It is not need consider “nir_probe_[2-10].txt” and “nir_probe_[2-10].txt” ?

3.  your paper "Dual Variational Generation for Low Shot Heterogeneous Face Recognition"  use LightCNN-29 or LightCNN-29v2 ?

**Comment by AlfredXiangWu** [comment]
1. Yes, the generator only uses "nir_train_1.txt" and "vis_train_1.txt".
2. The performance is tested on "nir_probe_1.txt" and "vis_gallery_1.txt". 
3. The pretrained face recognition model is LightCNN-29v2.

**Comment by ZYBOBO** [comment]
> 1. Get probe and gallery list.
> 2. Evaluate Rank-1 and ROC.
>    2.1 For rank-1 accuracy, you can select each probe image and compute the cosine similarity between it and all the gallery images.
>    2.2 For ROC, you can compute all the cosine similarity between probe and gallery (total about 200w+ pairs), and then feed them into a roc curve api such as sklearn.metrics.roc_curve.

Hello, from your reply to #5, I have some questions for your help. 
1. From your train_lightcnn.py, Rank-1 accuracy is not the cosine similarity beteween probe image and gallery image.   Is there anything different between train accuracy and validation accuracy?
2.  Do you mean that VR@FAR is the point from ROC? Can I regard VR@FAR as TPR@FPR? 
3. How do you determine the corresponding threshold when FAR is 1% or 0.1%?

I'm looking forward for your kind reply. Thanks~
